### ISSUE TITLE
Processing delay at prediction step

### ISSUE THREAD

**Issue by jbiggsets** [EducationalTestingService/rsmtool#123]
For a relatively large Train/Test set (300K + 100K, 44 feats), the `generate predicitions` step takes much longer than any other step.

 The delay is at the step predict:predict_with_the_model function before it starts post-processing the predictions.

**Comment by desilinguist** [comment]
So, I was looking into the issue that says the generating predictions is really slow for large datasets. I took the `train.csv` and `test.csv` files in our tests and replicated them 600x and 500x each to get a 300K training set and a 100K test set. I then annotated the code with timing commands for each of the 8 steps in `rsmtool` and ran the `lr` test 10 times. Here are the means for the times taken for each step:

```
loading_experiment_data       3.818596
pre_processing                0.162850
saving_train_test_to_disk    11.963902
train_model                   1.375136
run_analyses_train            1.760559
generate_predictions          9.526928
save_predictions_to_disk      3.297471
run_analyses_predictions     18.190961
report_generation             0.531004
```

So, prediction generation is not the _most_ expensive but among the steps that do not involve the disk, it's definitely up there. So, this is definitely still an issue.

**Comment by desilinguist** [comment]
So, I am looking into this a bit more. It looks most of the time in `generate_train_and_test_predictions()` is spent in the function `predict_with_model()`. I used [line_profiler](https://github.com/rkern/line_profiler) to profile the `predict_with_model()` function line by line and I get some very interesting results (see attached [file](https://github.com/EducationalTestingService/rsmtool/files/1682919/predict_with_model_lprof.txt)). The highlight is that 64% of the is spent on the following line: 

```
features = df[feature_columns].to_dict(orient='records')
```

And then another 35% is spent on actually creating the `FeatureSet` object:


```
fs = FeatureSet('data', ids=ids, labels=labels, features=features)
```

I guess the second line is something we can optimize in SKLL but we _should_ be able to find a faster way to create the feature dictionaries from pandas.

**Comment by desilinguist** [comment]
1. So, I tried an obvious alternative (using `iterrows()` on the data frame and calling `dict()` on each row separately) and as expected it was about 8x slower. 

2. After looking at the code for a bit, I realized that we _could_ probably make things a bit faster but at the cost of API "niceness". Right now, we pass in a dataframe containing the pre-processed features,  then we convert it to  list of dictionaries and then manually create a `FeatureSet` object. However, note that the pre-processed feature file is already written to disk, so we could basically just use a `CSVReader` and read the file from disk and that would directly give us a `FeatureSet` object. This would collapse the current two-step process of (1) converting a data frame to feature dictionaries and (2) creating a featureset object manually into a single step. However, I am not really convinced it would be that much faster since now the disk is involved. Even if it were faster, this would mean passing in the file paths instead of simply the data frames to `generate_train_and_test_predictions()` which doesn't seem like a great idea.

3. Looking at the implementation of `to_dict()` in `pandas`, it seems like it's fairly efficient. So, unless we want to write something in C, I am not sure we can do something that much better.

For now, I will close this issue as something we can't fix. We can revisit this later if it turns out to be a severe bottle neck.
